### ISSUE TITLE
Feature Request: encryption at rest

### ISSUE THREAD

**Issue by manuerwin** [scastiel/spliit2#34]
First of all, thanks so much for making this app <3

I would like my shared expense data to be encrypted when stored in the database (either via your hosted implementation or my own).

I’ve had a newb scan of the code and can’t tell if the data stored is encrypted?

I can certainly have a go at introducing this, however may take me awhile :)

**Comment by ChristopherJohnston** [comment]
if encryption/decryption is done at the client side then I think local storage of the key along with relevant encrypt/decrypt functions would be sufficient to satisfy this feature

**Comment by manuerwin** [comment]
Thanks @ChristopherJohnston for replying!
Like and agree with the client doing the heavy lifting re: encryption/decryption.
Do you have advice on how long the key should be retained in localstorage? or should it be in sessionstorage?

**Comment by scastiel** [comment]
It would be possible to end-to-end encrypt group names, participants names and expense titles, but that would mean having to enter a password the first time accessing the group. The password could be stored in local or session storage, and could even (optionally) be part of the shared URL.

With a different storage implementation, it would also be possible to encrypt *all* group information. Since there is no shared information between groups, it wouldn’t be for a group data to be stored in an (encrypted) JSON object. This implementation would require dealing with conflicts (maybe with CRDTs?) in case two people update the same group (e.g. add an expense at the same time).

**Comment by manuerwin** [comment]
Thanks Sebastien!

I was thinking the simpler version to start as it felt like group names, participants names and expense titles are the data that are “identifiable”. 

I’ll have a go at this unless someone beats me to it, may take a few weeks though :p

**Comment by ChristopherJohnston** [comment]
I had a little play around with the cryptography functions in libp2p, using a local IndexedDB to store keys

https://github.com/ChristopherJohnston/libp2p-keychain-encrypt-example

This could be a little bit too much, but might work as follows:

* On group creation, set a passphrase that would be shared with other members of the group.

* The passphrase and groupID are used by the createKey method in libp2p/keychain to create a private key, which is stored in the browser's IndexedDB.

* When creating the group link use the exportKey method in libp2p/keychain to export the private key from IndexedDB and and add it to the group link's querystring.

* When loading a group from the link, ask for the passphrase which is used with the private key in the querystring to pass into the importKey method of libp2p/keychain.

* This would create and store the key in IndexedDB. Good thing here is even with the group link, you need the passphrase to be able to see the data.

* When an expense is created, use the encrypt method in libp2p/cms to encrypt the individual values before sending to the server
* When an expense is received, use the decrypt method in libp2p/cms to decrypt the individual values before using them

Potential problems:

* If the private key and passphrase are lost, all data is lost

**Comment by ChristopherJohnston** [comment]
POC of the above here: https://github.com/scastiel/spliit2/pull/41

**Comment by scastiel** [comment]
Thanks @ChristopherJohnston I think your implementation makes total sense!

> If the private key and passphrase are lost, all data is lost

I’m especially worried about this point. I’m afraid it’s hard to let users know that they have to store or remember the passphrase, or the full shared link, somewhere, otherwise no one (even me) will be able to restore access to their group. Many people have trouble with this idea.

An alternative could be to let users decide wether or not they want to encrypt their data, with the advantages and drawback of both options. Again, it would require a bit of clear explanations, and to maintain two distinct logics.

**Comment by manuerwin** [comment]
Very much agree with making encryption optional with clear warnings/suggestions (e.g. use a password manager plus regularly backup transactions).

**Comment by Pt4r** [comment]
Are there any news regarding this issue?

**Comment by manuerwin** [comment]
I’ve not had a chance to look at this yet, hoping to do so next month.

**Comment by manuerwin** [comment]
@david-alm I'm a newb in this space, great to see minibone is built atop the [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) and agree symmetric key encryption will work well here.
Please please go ahead with your PR if you have time :)
If not, I'll continue my slow learning journey :p

**Comment by manuerwin** [comment]
> Just for my understanding, does Vercel Postgres not already encrypt data at rest?

@raymondji yes per https://vercel.com/docs/security/compliance#data-encryption which is great, however does mean whoever has the keys can decrypt ALL expenses/groups/etc stored, whereas I'm thinking about effectively enabling end-to-end encryption per group.

**Comment by erikbrinkman** [comment]
reviving this thread to gauge what the desire is for this feature. I have a POC that seems to be roughly working where on group creation it generates a private key. All fields (except mtime) are then nulled in the database and instead an encrypted blob is stored in a new column. The privat ekey is then shared a long with the link. Like most other groups this lives primarily in local memory with the abiltiy to link share. the POC has it opt in at group creation but otherwise not editable. I think this approach solves many of the issues proposed already, although obviously there's no free lunch. Is this still something that's desired? What's the beat way to hammer out a design?
